### ISSUE TITLE
ShopStrategy#update_access_scopes? should only be true when user needs to approve new scopes

### ISSUE THREAD

**Issue by zernie** [Shopify/shopify_app#1820]
# Issue summary

Before opening this issue, I have:

- [x] Upgraded to the latest version of the package
  - `shopify_app` version:
  - Ruby version: 3.2.0
  - Operating system: Ubuntu
- [ ] Set `log_level: :debug` [in my configuration](https://github.com/Shopify/shopify-api-ruby#setup-shopify-context), if applicable
- [x] Found a reliable way to reproduce the problem that indicates it's a problem with the package
- [x] Looked for similar issues in this repository
- [x] Checked that this isn't an issue with a Shopify API
  - If it is, please create a post in the [Shopify community forums](https://community.shopify.com/c/partners-and-developers/ct-p/appdev) or report it to [Shopify Partner Support](https://help.shopify.com/en/support/partners/org-select)

We are dynamically requesting access scopes based on what features our stores require. This means that  `config.scope = ''`.

When a store requests a certain feature, we prompt him to go through the OAuth flow with the additional scopes.  This works fine.

## Expected behavior

Access scopes should not change.

## Actual behavior

When a user logs out & signs in again, the shop access scopes are reset. The reason for that is this method:

```ruby
module ShopifyApp
  module AccessScopes
    class ShopStrategy
      class << self
        def update_access_scopes?(shop_domain)
          shop_access_scopes = shop_access_scopes(shop_domain)
          configuration_access_scopes != shop_access_scopes
        end
```

However, it should only return true if the shop is missing some scopes, not when it has more that are set in `config.scope`. I believe `UserStrategy` is already following this logic.

**Comment by Eianaftyio** [comment]
config.scope

**Comment by paulomarg** [comment]
Hey, thanks for raising this. I think this makes sense, we'll look into it and make sure the suggested fix doesn't carry any risks.

**Comment by zernie** [comment]
Hey @paulomarg thanks a lot for the quick response 

> Hey, thanks for raising this. I think this makes sense, we'll look into it and make sure the suggested fix doesn't carry any risks - **I think we might have to replace the token if you remove scopes from an app via the config value, which might not play well with the fix.**

I'm not quite sure I get this.

In our case, `config.scope` never changes. 

However, when a certain user needs an additional scope he is prompted to the confirmation screen (/login?...) with the missing scopes. This works great, (although it'd be great if this flow was documented)`shop.access_scopes` are updated properly and an offline `shopify_token` is saved. Here's how we do it:

```ruby 
callback_url = ShopifyApp.configuration.login_callback_url.gsub(%r{^/}, '')
        cookie, auth_route = ShopifyAPI::Auth::Oauth.begin_auth(
          shop: current_shop.shopify_domain,
          redirect_path: "/#{callback_url}",
          is_online: false,
          scope_override: scopes
        ).values_at(:cookie, :auth_route)

        cookies.encrypted[cookie.name] = {
          expires: cookie.expires,
          secure: true,
          http_only: true,
          value: cookie.value
        }

        auth_route
```

However, after a re-login with `config.reauth...` enabled, `shop.access_scopes` are reset back to `config.scope`, which is empty.

And we also need to re-request an online token for the user.

**Comment by paulomarg** [comment]
> > Hey, thanks for raising this. I think this makes sense, we'll look into it and make sure the suggested fix doesn't carry any risks - I think we might have to replace the token if you remove scopes from an app via the config value, which might not play well with the fix.

> I'm not quite sure I get this.

I was mostly talking about other scenarios, where the app uses config.scopes but removes one of those - I think we might run into trouble in that scenario, but it doesn't apply here.

In any case, I think I understand your problem better now - basically, your override is working for the first auth, but it doesn't get re-applied because [we default to `config.scopes` in those cases](https://github.com/Shopify/shopify_app/blob/main/app/controllers/shopify_app/sessions_controller.rb#L72-L92).

I'm assuming you're already overriding the sessions controller in order to set your overrides? If not, that might be an alternative.

I also noticed that there is currently no way to override [the shop scope strategy](https://github.com/Shopify/shopify_app/blob/main/lib/shopify_app/configuration.rb#L84-L88) in the config, other than monkeypatching. One way to solve this would be to override the scope strategy (i.e. write your own version of [`ShopStrategy`](https://github.com/Shopify/shopify_app/blob/main/lib/shopify_app/access_scopes/shop_strategy.rb)). That way you'd have full control over how that comparison works.

WDYT?

**Comment by zernie** [comment]
> I'm assuming you're already overriding the sessions controller in order to set your overrides? If not, that might be an alternative.

nope, we haven't considered that. we have a separate controller that redirects user to the `/login` page.


> I also noticed that there is currently no way to override [the shop scope strategy](https://github.com/Shopify/shopify_app/blob/main/lib/shopify_app/configuration.rb#L84-L88) in the config, other than monkeypatching. One way to solve this would be to override the scope strategy (i.e. write your own version of [ShopStrategy](https://github.com/Shopify/shopify_app/blob/main/lib/shopify_app/access_scopes/shop_strategy.rb)) and configure that as an override. That way you'd have full control over how that comparison works.

Yep, that'd be great! Currently we have to resort to monkey-patching.


---

There's one more issue that we've noticed. Sometimes `app/uninstalled` webhook is delayed by a couple of minutes and we can't reset `Shop#access_scopes` in time.

In addition to that, a webhook job can be slow to process. Have you considered using `ShopifyAPI::AccessScope.all` instead of `Shop#access_scopes` to check whether access scopes need to be requested? This should fix all potential issues.

**Comment by paulomarg** [comment]
> nope, we haven't considered that. we have a separate controller that redirects user to the /login page when a customer needs new scopes.

That's fair - I think for this to work properly with the current code, the right fix would be to replace the sessions controller with one that extends [the default](https://github.com/Shopify/shopify_app/blob/main/app/controllers/shopify_app/sessions_controller.rb#L72-L92), and replace the `redirect_to_begin_oauth` method to also include your scopes - that way, automatic re-auths will also request the appropriate scopes.

For the future, I feel like we're lacking the ability to customize this flow - maybe a `custom_shop_scopes` setting that can be a callback would do the trick. That way you wouldn't even need a custom login flow to begin with, you'd just add that config and we'd do the authing for you.

For the sake of transparency, assuming that overriding the sessions controller works, there are other things that might take precedence over this, but I'll add it to our tracking.

> Have you considered using ShopifyAPI::AccessScope.all instead of Shop#access_scopes to check whether access scopes need to be requested?

Yes, webhooks can take a little while to be sent. The big downside of getting the scopes from the API is that we basically run this check on every request, which means an added round trip to Shopify. That might slow apps down too much, so I'm not sure that's something we want to do.
